pump/server_test.go

```diff
@@ -555,3 +562,274 @@ func (s *listenSuite) TestReturnListener(c *C) {
+}
+
+func (s *newServerSuite) TestCreateNewPumpServer(c *C) {
+	getPdClient = func(string, security.Config) (pd.Client, error) {
+		return &mockPdCli{}, nil
+	}
+	newTiKVLockResolver = func([]string, config.Security) (*tikv.LockResolver, error) {
+		return nil, nil
+	}
+	newKVStore = func(path string) (kv.Storage, error) {
+		return nil, nil
+	}
+
+	p, err := NewServer(s.cfg)
+	c.Assert(err, IsNil)
+	c.Assert(p.clusterID, Equals, uint64(8012))
+}
+
+type startNode struct {
+	status *node.Status
+}
+
+func (n *startNode) ID() string                                                   { return "startnode-long" }
+func (n *startNode) ShortID() string                                              { return "startnode" }
+func (n *startNode) RefreshStatus(ctx context.Context, status *node.Status) error { return nil }
+func (n *startNode) Heartbeat(ctx context.Context) <-chan error                   { return make(chan error) }
+func (n *startNode) Notify(ctx context.Context) error                             { return nil }
+func (n *startNode) NodeStatus() *node.Status {
+	return n.status
+}
+func (n *startNode) NodesStatus(ctx context.Context) ([]*node.Status, error) {
+	return []*node.Status{n.status}, nil
+}
+func (n *startNode) Quit() error { return nil }
+
+type startStorage struct {
+	sig chan struct{} // use sig to notify that the closing work has been done
+}
+
+func (s *startStorage) AllMatched() bool                            { return true }
+func (s *startStorage) WriteBinlog(binlogItem *binlog.Binlog) error { return nil }
+func (s *startStorage) GetGCTS() int64                              { return 0 }
+func (s *startStorage) GC(ts int64)                                 {}
+func (s *startStorage) MaxCommitTS() int64                          { return 0 }
+func (s *startStorage) GetBinlog(ts int64) (*binlog.Binlog, error) {
+	return nil, errors.New("server_test")
+}
+func (s *startStorage) PullCommitBinlog(ctx context.Context, last int64) <-chan []byte {
+	return make(chan []byte)
+}
+func (s *startStorage) Close() error {
+	<-s.sig
+	// wait for pump server to change node back to startNode, or test etcd server might be closed
+	time.Sleep(20 * time.Microsecond)
+	return nil
+}
+
+type startServerSuite struct {
+	origUtilGetTSO func(pd.Client) (int64, error)
+}
+
+var _ = Suite(&startServerSuite{})
+
+func (s *startServerSuite) SetUpTest(c *C) {
+	s.origUtilGetTSO = utilGetTSO
+}
+
+func (s *startServerSuite) TearDownTest(c *C) {
+	utilGetTSO = s.origUtilGetTSO
+}
+
+func (s *startServerSuite) TestStartPumpServer(c *C) {
+	utilGetTSO = func(pd.Client) (int64, error) {
+		return 0, nil
+	}
+	etcdClient := testEtcdCluster.RandClient()
+	cfg := &Config{
+		ListenAddr:        "http://127.0.0.1:8250",
+		AdvertiseAddr:     "http://127.0.0.1:8260",
+		Socket:            "unix://127.0.0.1:" + string(time.Now().UnixNano()) + "/hello/world",
+		EtcdURLs:          strings.Join(etcdClient.Endpoints(), ","),
+		DataDir:           "/tmp/pump",
```

Comment:
os.TempDir

Reply:
I have switched this to `path.Join(c.MkDir(), "pump")`.